sql/expression/function/logarithm.go

```diff
@@ -0,0 +1,223 @@
+	b := base
+	return func(e sql.Expression) sql.Expression {
+		return NewLogBase(b, e)
+	}
+}
+
+// LogBase is a function that returns the logarithm of a value with a specific base.
+type LogBase struct {
+	expression.UnaryExpression
+	base float64
+}
+
+// NewLogBase creates a new LogBase expression.
+func NewLogBase(base float64, e sql.Expression) sql.Expression {
+	return &LogBase{UnaryExpression: expression.UnaryExpression{Child: e}, base: base}
+}
+
+func (l *LogBase) String() string {
+	switch l.base {
+	case float64(math.E):
+		return fmt.Sprintf("ln(%s)", l.Child)
+	case float64(10):
+		return fmt.Sprintf("log10(%s)", l.Child)
+	case float64(2):
+		return fmt.Sprintf("log2(%s)", l.Child)
+	default:
+		return fmt.Sprintf("log(%v, %s)", l.base, l.Child)
+	}
+}
+
+// TransformUp implements the Expression interface.
+func (l *LogBase) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	child, err := l.Child.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	return f(NewLogBase(l.base, child))
+}
+
+// Type returns the resultant type of the function.
+func (l *LogBase) Type() sql.Type {
+	return sql.Float64
+}
+
+// IsNullable implements the sql.Expression interface.
+func (l *LogBase) IsNullable() bool {
+	return l.Child.IsNullable()
+}
+
+// Eval implements the Expression interface.
+func (l *LogBase) Eval(
+	ctx *sql.Context,
+	row sql.Row,
+) (interface{}, error) {
+	v, err := l.Child.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if v == nil {
+		return nil, nil
+	}
+
+	val, err := sql.Float64.Convert(v)
+	if err != nil {
+		return nil, sql.ErrInvalidType.New(reflect.TypeOf(v))
+	}
+	return computeLog(val.(float64), l.base)
+}
+
+// Log is a function that returns the natural logarithm of a value.
+type Log struct {
+	expression.BinaryExpression
+}
+
+// NewLn creates a new Log expression.
+func NewLog(args ...sql.Expression) (sql.Expression, error) {
+	argLen := len(args)
+	if argLen == 0 || argLen > 2 {
+		return nil, sql.ErrInvalidArgumentNumber.New("1 or 2", argLen)
+	}
+
+	var right sql.Expression
+	if len(args) == 2 {
+		right = args[1]
+	}
+
+	return &Log{expression.BinaryExpression{Left: args[0], Right: right}}, nil
+}
+
+func (l *Log) String() string {
+	if l.Right == nil {
+		return fmt.Sprintf("log(%s)", l.Left)
+	}
+
+	return fmt.Sprintf("log(%s, %s)", l.Left, l.Right)
+}
+
+// TransformUp implements the Expression interface.
+func (l *Log) TransformUp(f sql.TransformExprFunc) (sql.Expression, error) {
+	var args = make([]sql.Expression, 2)
+	arg, err := l.Left.TransformUp(f)
+	if err != nil {
+		return nil, err
+	}
+	args[0] = arg
+
+	args[1] = nil
+	if l.Right != nil {
+		arg, err := l.Right.TransformUp(f)
+		if err != nil {
+			return nil, err
+		}
+		args[1] = arg
+	}
+
+	expr, err := NewLog(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	return f(expr)
+}
+
+// Children implements the Expression interface.
+func (l *Log) Children() []sql.Expression {
+	if l.Right == nil {
```

Comment:
IMO internally we should explicitly say - left is a base, right is argument.
If base is not specified in constructor then we set it up to `E`, but internally we shouldn't have hidden logic, which argument stands for what.

Reply:
If we setup base to Math.E wouldn't that make difficult to detect if the request was using 1 argument instead of 2. For example when l.Right == nil we know that the request was `log(%s)` but now every time it would be `log(%s, %s)` which looks wierd.

```
mysql> select log(2);
+---------------------------+
| log(2.718281828459045, 2) |
+---------------------------+
|        0.6931471805599453 |
+---------------------------+
```

Reply:
But the header comes from the `String` - right?
If yes, maybe we can add extra nice formatting

Reply:
Ok we can just always show the base even if its not specified.